examples/plugins/local/my-plugin/my-plugin.json

```diff
@@ -0,0 +1,13 @@
+{
+  "name": "my-plugin",
+  "version": "0.0.1",
+  "readme": "Example custom plugin",
+  "env": {
+    "MY_FOO_VAR": "BAR"
+  },
+  "create_files": {
+    "{{ .Virtenv }}/empty-dir": "",
+    "{{ .Virtenv }}/some-file": "some-file.txt",
+    "{{ .Virtenv }}/process-compose.yaml": "process-compose.yaml"
```

Comment:
I'm wondering if we can drop the `{{ .Virtenv }}`? Why do we need it?

I feel a more natural way for a user to write this is:
```
"create_files": {
   "empty-dir": "",
   "some-file": "some-file.txt",
 }
```

Reply:
To clarify: I fear its easy for the user to make a mistake with the {{ .Virtenv }} format. 

If we can help it, it would be nice to enable a genuine json which is not also a go-template.

Reply:
Agree that having pure json would be nicer, but there are actually 2 different directories (Virtenv and devbox.d) that plugins write to. It would feel a bit weird to default to Virtenv and then require a template for devbox.d.

Are template variables that high a bar?

Reply:
ohh I see. 

My concern about the template variables is that its a golang-ism that I fear (fairly strongly) will confuse users who are not familiar with golang (which I'd guess is a fair %, especially if devbox gains more users).

If there are strictly two (or a finite set of) options, then how about making that part of the spec. Something like:
```
"create_files": {
    "virtenv": {
     "empty-dir": "",
     "some-file": "some-file.txt",
   },
   "devbox.d": {
      "process-config": "process-conf.yaml"
   }
 },
```
Thoughts? In our docs and examples, we can show that `virtenv` and `devbox.d` are the only two acceptable keys.

Alternate names for the keys could be:
-  `internal` and `external`, so that plugin authors know which files are internal to their plugin, versus externally editable by users.
- `plugin-owned` and `user-owned`.

Reply:
There are quite a few template variables and they provide a pretty large amount of customizability so I would prefer to always allow them (even if we allow shortcuts or syntax sugar).


re: template variables being a golang-ism: this is true, but they are fairly common. (See helm for example).

I think your suggestion of having special keys in `create_files` is good. My main counter is I would still allow using template variables and custom paths to allow more customizability. And a (better?) possible replacement is https://github.com/jetpack-io/tyson and if we want to do that, we might regret adding special fields to `create_files`.

Thoughts?

Reply:
Wait, what are the other template-variables? I'm feeling pretty uncomfortable about exposing our internal plugin schema directly for external usage where we'll have to support it indefinitely. 

Could we write down a spec. for this plugin's schema, so we can evaluate it independent of the implementation in this PR?

> And a (better?) possible replacement is https://github.com/jetpack-io/tyson and if we want to do that, we might regret adding special fields to create_files.

Can you elaborate on this: why would using TySON be harder with a schema that has whitelisted keys for `create_files`? How is it different from the whitelisted keys at the top-level like `version` and `env`?

Reply:
> Wait, what are the other template-variables?

plugins variables are defined [here](https://github.com/jetpack-io/devbox/blob/bec133a3306a357e1ec51eb0f52fa346f8f79210/internal/plugin/plugin.go#L238-L242)

Created file variables are defined [here](https://github.com/jetpack-io/devbox/blob/bec133a3306a357e1ec51eb0f52fa346f8f79210/internal/plugin/plugin.go#L163-L171)

> I'm feeling pretty uncomfortable about exposing our internal plugin schema directly for external usage where we'll have to support it indefinitely.

Any syntax (template variables or otherwise) we'll have to maintain "forever" (in quotes because nothing is really forever, we can always force people to update). e.g. if you add special create_file fields you are essentially saying you will maintain that forever.

 > Could we write down a spec.

For expediency, my hope was to use our current variable set and document them as experimental/may change until 1.0.0. We obviously do not want them to change, but it seems hard to guarantee. I'm OK with documenting a subset and saying everything else is private and you should not use. 

> Can you elaborate on this: why would using TySON be harder with a schema that has whitelisted keys for create_files?

I think you misunderstood me. With TySON (and a simple API for plugins) we no longer need templates or special keys.  You could do:

```ts
import { virtenvPath } from '@devbox/plugins';

export default {
  name: "foo",
  create_files: {
    [virtenvPath() + "/my-file.txt"]: "my-file.txt",
    [virtenvPath() + "/other.txt"]: "other.txt"
  }
}
```

Since the ts API is published and versioned, we could make somewhat breaking changes without breaking people's plugins.